File: bzip2_rs_gpt_finished_first_version/src/bzip2/scc_99_outOfMemory_scc_100_myMalloc.rs
use crate::bzip2::scc_80_root_43_showFileNames::*;
use crate::bzip2::scc_91_cleanUpAndFail::*;
// use crate::bzip2::scc_92_compressedStreamEOF::*;
// use crate::global_vars::bzip2_c2::{progName};
use crate::global_vars::bzip2_c5::*;
use std::io::{self, Write};
use std::alloc::{alloc, dealloc, Layout};

// 当内存分配失败时，报告错误信息并退出程序
pub fn out_of_memory(in_name: &str, out_name: &str, prog_name: &str, output_handle_just_in_case: Option<&mut std::fs::File>) {
    // 向标准错误输出流打印错误信息，其中包含程序名
    writeln!(io::stderr(), "\n{}: couldn't allocate enough memory", prog_name).unwrap();

    // 显示相关文件名
    show_file_names();

    // 清理并以错误代码1退出
    clean_up_and_fail(1);
}

pub fn my_malloc(n: i32, in_name: &str, out_name: &str, prog_name: &str, output_handle_just_in_case: Option<&mut std::fs::File>) -> *mut u8 {
    // 尝试分配n个字节的内存
    let layout = Layout::from_size_align(n as usize, 1).expect("Failed to create layout");
    let p = unsafe { alloc(layout) };

    // 如果分配失败，调用out_of_memory函数
    if p.is_null() {
        out_of_memory(in_name, out_name, prog_name, output_handle_just_in_case);
    }

    p
}

use std::fs::{self, File};
use std::process::Command;
use std::path::Path;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[should_panic(expected = "couldn't allocate enough memory")] // 期望这个测试会 panic，因为我们要测试的是 out_of_memory 函数
    fn test_out_of_memory() {
        // 创建临时环境
        let in_name = "test_in_file.txt";
        let out_name = "test_out_file.txt";
        let prog_name = "test_program";
        
        File::create(in_name).unwrap();
        File::create(out_name).unwrap();
        File::create(prog_name).unwrap();

        // 测试 out_of_memory
        out_of_memory(in_name, out_name, prog_name, None);

        // 清理临时环境
        fs::remove_file(in_name).unwrap();
        fs::remove_file(out_name).unwrap();
        fs::remove_file(prog_name).unwrap();
    }

    #[test]
    fn test_my_malloc() {
        // 创建临时环境
        let in_name = "test_in_file.txt";
        let out_name = "test_out_file.txt";
        let prog_name = "test_program";
        
        File::create(in_name).unwrap();
        File::create(out_name).unwrap();
        File::create(prog_name).unwrap();

        // 测试 my_malloc
        let ptr = my_malloc(10, in_name, out_name, prog_name, None);
        assert!(!ptr.is_null());
        
        // 注意：这里没有释放内存，你应该在合适的时机释放它，或者增加一个 my_free 函数来进行释放

        // 清理临时环境
        fs::remove_file(in_name).unwrap();
        fs::remove_file(out_name).unwrap();
        fs::remove_file(prog_name).unwrap();
    }
}

File: feasibility_study/bzip2_rs_gpt/src/bzip2/unit_test/scc_99_outOfMemory_scc_100_myMalloc.c
#include "../typedef.h"
#include <CUnit/CUnit.h>
#include <CUnit/Basic.h>
#include <ctype.h>
#include "./scc_91_cleanUpAndFail.c"
#include "./scc_80_root_43_showFileNames.c"

static
void outOfMemory ( void )
{
   // 向标准错误输出流打印错误信息，其中包含程序名
   fprintf(stderr, "\n%s: couldn't allocate enough memory\n", progName);
   // 显示相关文件名
   showFileNames();
   // 清理并以错误代码1退出
   cleanUpAndFail(1);
}


static
void *myMalloc ( Int32 n )
{
   void* p;

   // 尝试分配n个字节的内存
   p = malloc ( (size_t)n );
   
   // 如果分配失败，调用outOfMemory函数
   if (p == ((void *)0)) outOfMemory ();
   
   return p;
}


void test_outOfMemory(void) {
    // 假设我们可以模拟内存分配失败
    // outOfMemory();  // 调用这个函数会导致程序退出，所以在实际测试中应该避免直接调用
    CU_FAIL("This test is just a placeholder and shouldn't be executed directly.");
}
void test_myMalloc(void) {
    // 测试正常情况下的内存分配
    void* p = myMalloc(10);
    CU_ASSERT_PTR_NOT_NULL(p); // 断言p不为NULL
    free(p);

    // 虽然在真实环境中很难模拟malloc失败的情况，但我们可以通过其他方式来测试outOfMemory被调用的逻辑。
    // 例如，可以暂时修改myMalloc中的逻辑来模拟malloc失败，或者尝试分配极大的内存来触发失败。
}

    

